ARC-0200 draft (#223)
algorandfoundation/ARCs: ARCs/arc-0200.md

However, to provide rich extra functionality, it can be desirable to implement tokens as a smart contract instead.
To foster an interoperable token ecosystem, it is necessary that the core interfaces for tokens be standardized.

## Specification
The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.

### Core Token specification

A smart contract token that is compliant with this standard MUST implement the following interface:

```json
{
  "name": "ARC-200",
  "desc": "Smart Contract Token Base Interface",
  "methods": [
    {
      "name": "name",
      "desc": "Returns the name of the token",
      "readonly": true,
      "args": [],
      "returns": { "type": "bytes[32]", "desc": "The name of the token" }
    },
    {
      "name": "symbol",
      "desc": "Returns the symbol of the token",
      "readonly": true,
      "args": [],
      "returns": { "type": "bytes[8]", "desc": "The symbol of the token" }
    },
    {

>       "name": "decimals",

Author: The reason for treating `decimals` as an individual field is similar to how `name` or `symbol` are handled as separate fields. I prefer not to use `ufixed` encoded decimal because it would limit the implementation to a specific encoding method. The ABI specification only requires `decimals` to be returned as a uint8, but it allows the smart contract to store the decimal value in any desired format. The choice of implementation is left to the developer.

One possible implementation approach is to validate that `decimals` falls within an acceptable range. For compatibility with ASAs, the decimals value could be checked to be less than 19. For ERC20 compatibility, an implementation would check that it is less than 256.

Additionally, based on my understanding, Reach does not utilize `ufixed`. If a value is known to be less than 256 at compile-time, the compiler would automatically use uint8.

Considering the potential inconveniences and the availability of alternative methods, I don't believe using `ufixed` is necessary to achieve the desired purpose.

>       "name": "decimals",

Reviewer: Yeah the more I think about it the more it makes sense to keep it as a separate method since clients will need an easy way to discover how many decimals are actually supported regardless of encoding. 

I will say though that I think the only language that should dictate the direction of standards is TEAL. Standards should dictate language features, not the other way around. Obviously it is extra dev effort and will take longer to implement/support across the ecosystem. But long term, having a standard that everyone feels is best suited for its purpose is better than having multiple evolving standards superseding each other.

>       "name": "decimals",

Author: Here since this values may use UInt256, `uint8` may be preferable `decimals`

>       "name": "decimals",

Author: > Yeah the more I think about it the more it makes sense to keep it as a separate method since clients will need an easy way to discover how many decimals are actually supported regardless of encoding.

Yup. It is an individual field. Disregard my tangent regarding decimals and uint8. Similar to using uint256 for totalSupply, balances, and allowance values, the decimals method should return a uint8.

The rest are implementation details which I hope we all agree is out of the scope of the spec.

This ARC (Algorand Request for Comments) specifies an interface for tokens to be implemented on Algorand as smart contracts. The interface defines a minimal interface required for tokens to be held and transferred, with the potential for further augmentation through additional standard interfaces and custom methods.

## Motivation

Currently, most tokens in the Algorand ecosystem are represented by ASAs (Algorand Standard Assets). However, to provide rich extra functionality, it can be desirable to implement tokens as smart contracts instead. To foster an interoperable token ecosystem, it is necessary that the core interfaces for tokens be standardized.

## Specification

The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.

### Core Token specification

A smart contract token that is compliant with this standard MUST implement the following interface:

```json
{
  "name": "ARC-200",
  "desc": "Smart Contract Token Base Interface",
  "methods": [
    {

>       "name": "name",

Reviewer 2: ```suggestion
      "name": "arc200_name",
```

      "desc": "Returns the name of the token",
      "readonly": true,
      "args": [],
      "returns": { "type": "bytes[32]", "desc": "The name of the token" }
    },
    {

>       "name": "symbol",

Reviewer 2: ```suggestion
      "name": "arc200_symbol",
```

      "desc": "Returns the symbol of the token",
      "readonly": true,
      "args": [],
      "returns": { "type": "bytes[8]", "desc": "The symbol of the token" }
    },
    {

>       "name": "decimals",

Reviewer 2: ```suggestion
      "name": "arc200_decimals",
```

      "desc": "Returns the decimals of the token",
      "readonly": true,
      "args": [],
      "returns": { "type": "uint8", "desc": "The decimals of the token" }
    },
    {

>       "name": "totalSupply",

Reviewer 2: ```suggestion
      "name": "arc200_totalSupply",
```

      "desc": "Returns the total supply of the token",
      "readonly": true,
      "args": [],
      "returns": { "type": "uint256", "desc": "The total supply of the token" }
    },
    {

>       "name": "balanceOf",

Reviewer 2: ```suggestion
      "name": "arc200_balanceOf",
```

      "desc": "Returns the current balance of the owner of the token",
      "readonly": true,
      "args": [
        {
          "type": "address",
          "name": "owner",
          "desc": "The address of the owner of the token"
        }
      ],
      "returns": {
        "type": "uint256",
        "desc": "The current balance of the holder of the token"
      }
    },
    {

>       "name": "transfer",

Reviewer 2: ```suggestion
      "name": "arc200_transfer",
```

      "desc": "Transfers tokens",
      "readonly": false,
      "args": [
        {
          "type": "address",
          "name": "to",
          "desc": "The destination of the transfer"
        },
        {
          "type": "uint256",
          "name": "value",
          "desc": "Amount of tokens to transfer"
        }
      ],

>       "returns": { "type": "bool", "desc": "Success" }

Reviewer 3: Do you see any advantage in making `transfer` and `transferFrom` return void instead of booleans? My impression is that for ERC-20 implementations, the safer thing to do is to only ever return true or error if the operation isn't possible. Otherwise it's very important that callers check the returned value, which is just another opportunity for a bug.

Author: It does trivially return True. To be on the safe side you would check that it does despite the fact that in practice the only alternative is to throw an error. 
I think there is no advantage of returning void instead of boolean. Same problem more uncertainty.
I know we are not going into implementation but if we were I would advise having both methods return true if not error and don't even bother having code returning false.

Author: In other words, it is not written in the spec but these methods only return true, which is a bool

Reviewer 3: Then why not write into the spec that the methods should return only true? Or better yet, remove all uncertainty and have it return void?

True, this would be a departure from the ERC-20 spec, but a very small one that's easily bridged.

Reviewer: I think the argument for `bool` is that one can attempt to do a transfer which is a feature that isn't otherwise possible. The counterargument is that one should just check `allowance` before attempting a transfer.

Author: I think bool is fine the way it is. I don't think we should bake into the spec the fact that on the underlying network if something does not match up, assertions fail, it will crash and burn.

    },
    {

>       "name": "transferFrom",

Reviewer 2: ```suggestion
      "name": "arc200_transferFrom",
```

      "desc": "Transfers tokens from source to destination as approved spender",
      "readonly": false,
      "args": [
        {
          "type": "address",
          "name": "from",
          "desc": "The source  of the transfer"
        },
        {
          "type": "address",
          "name": "to",
          "desc": "The destination of the transfer"
        },
        {
          "type": "uint256",
          "name": "value",
          "desc": "Amount of tokens to transfer"
        }
      ],
      "returns": { "type": "bool", "desc": "Success" }
    },
    {

>       "name": "approve",

Reviewer 2: ```suggestion
      "name": "arc200_approve",
```

      "desc": "Approve spender for a token",
      "readonly": false,
      "args": [
        { "type": "address", "name": "spender" },
        { "type": "uint256", "name": "value" }
      ],
      "returns": { "type": "bool", "desc": "Success" }
    },
    {

>       "name": "allowance",

Reviewer 2: ```suggestion
      "name": "arc200_allowance",
```

      "desc": "Returns the current allowance of the spender of the tokens of the owner",
      "readonly": true,
      "args": [
        { "type": "address", "name": "owner" },
        { "type": "address", "name": "spender" }
      ],
      "returns": { "type": "uint256", "desc": "The remaining allowance" }
    }
  ],
  "events": [
    {

>       "name": "Transfer",

Reviewer 2: ```suggestion
      "name": "arc200_Transfer",
```

      "desc": "Transfer of tokens",
      "args": [
        {
          "type": "address",
          "name": "from",
          "desc": "The source of transfer of tokens"
        },
        {
          "type": "address",
          "name": "to",
          "desc": "The destination of transfer of tokens"
        },
        {
          "type": "uint256",
          "name": "value",
          "desc": "The amount of tokens transferred"
        }
      ]
    },
    {

>       "name": "Approval",

Reviewer 2: ```suggestion
      "name": "arc200_Approval",
```

      "desc": "Approval of tokens",
      "args": [
        {
          "type": "address",
          "name": "owner",
          "desc": "The owner of the tokens"
        },
        {
          "type": "address",
          "name": "spender",
          "desc": "The approved spender of tokens"
        },
        {
          "type": "uint256",
          "name": "value",
          "desc": "The amount of tokens approve"
        }
      ]
    }
  ]
}
```

Ownership of a token by a zero address indicates that a token is out of circulation indefinitely, or otherwise burned or destroyed.

> The methods `transfer` and `transferFrom` method MUST error when the balance of `from` is insufficient. In the case of the `transfer` method, from is implied as the `owner` of the token.

Reviewer 2: ```suggestion
The methods `arc200_transfer` and `arc200_transferFrom` method MUST error when the balance of `from` is insufficient. In the case of the `arc200_transfer` method, from is implied as the `owner` of the token.
```

> The `transferFrom` method MUST error unless called by an `spender` approved by an `owner`.

Reviewer 2: ```suggestion
The `arc200_transferFrom` method MUST error unless called by an `spender` approved by an `owner`.
```

> The methods `transfer` and `transferFrom` MUST emit a `Transfer` event.

Reviewer 2: ```suggestion
The methods `arc200_transfer` and `arc200_transferFrom` MUST emit a `Transfer` event.
```

> A `Transfer` event SHOULD be emitted, with `from` being the zero address, when a token is minted.

Reviewer 2: ```suggestion
A `arc200_Transfer` event SHOULD be emitted, with `from` being the zero address, when a token is minted.
```

> A `Transfer` event SHOULD be emitted, with `to` being the zero address, when a token is destroyed.

Reviewer 2: ```suggestion
A `arc200_Transfer` event SHOULD be emitted, with `to` being the zero address, when a token is destroyed.
```

> The `Approval` event MUST be emitted when an `approve` or `transferFrom` method is called successfully.

Reviewer 2: ```suggestion
The `arc200_Approval` event MUST be emitted when an `arc200_approve` or `arc200_transferFrom` method is called successfully.
```

> A value of zero for the `approve` method and the `Approval` event indicates no approval.

Reviewer 2: ```suggestion
A value of zero for the `arc200_approve` method and the `arc200_Approval` event indicates no approval.
```

> The `transferFrom` method and the `Approval` event indicates the approval value after it is decremented.

Reviewer 2: ```suggestion
The `arc200_transferFrom` method and the `arc200_Approval` event indicates the approval value after it is decremented.
```

---
arc: 200
title: Algorand Smart Contract Token Specification
description: Base specification for tokens implemented as smart contracts
author: Nicholas Shellabarger (@temptemp3)
discussions-to: https://github.com/algorandfoundation/ARCs/issues/223
status: Draft
type: Standards Track
category: Interface
created: 2023-07-03
requires: 3, 4, 22, 28
---

# Algorand Smart Contract Token Specification

## Abstract

This ARC (Algorand Request for Comments) specifies an interface for tokens to be implemented on Algorand as smart contracts. The interface defines a minimal interface required for tokens to be held and transferred, with the potential for further augmentation through additional standard interfaces and custom methods.

## Motivation

> Currently, most tokens in the Algorand ecosystem are represented by ASAs (Algorand Standard Assets). However, to provide rich extra functionality, it can be desirable implementing tokens as a smart contract instead. To foster an interoperable token ecosystem, it is necessary that the core interfaces for tokens be standardized.

Reviewer 4: ```suggestion
Currently, most tokens in the Algorand ecosystem are represented by ASAs (Algorand Standard Assets). However, to provide rich extra functionality, it can be desirable to implement tokens as smart contracts instead. To foster an interoperable token ecosystem, it is necessary that the core interfaces for tokens be standardized.
```

        { "type": "address", "name": "owner_" },
        { "type": "address", "name": "spender_" }
      ],
      "returns": { "type": "uint256", "desc": "The remaining allowance" }
    }
  ],
  "events": [
    {
      "name": "Transfer",
      "desc": "Transfer of tokens",
      "args": [
        {
          "type": "address",
          "name": "from_",
          "desc": "The source of transfer of tokens"
        },
        {
          "type": "address",
          "name": "to_",
          "desc": "The destination of transfer of tokens"
        },
        {
          "type": "uint256",
          "name": "value_",
          "desc": "The amount of tokens transferred"
        }
      ]
    },
    {
      "name": "Approval",
      "desc": "Approval of tokens",
      "args": [
        {
          "type": "address",
          "name": "owner_",
          "desc": "The owner of the tokens"
        },
        {
          "type": "address",
          "name": "spender_",
          "desc": "The approved spender of tokens"
        },
        {
          "type": "uint256",
          "name": "value_",
          "desc": "The amount of tokens approve"
        }
      ]
    }
  ]
}
```

> Ownership of a token by a zero address indicates that a token is out of circulation indefinitely, or otherwise burned or destroyed.

Reviewer 4: Is there a requirement that `totalSupply` reflect this convention?  Should `totalSupply` decrease after a transfer to the ZeroAddress?

Author: Currently no. In my mind, `totalSupply` is immutable. That is total supply starts at the mint address. Later on it may find itself in the zero address. However, depending on how you look at it circulating supply may differ. Circulating supply could be:

1) Tokens outside of mint address and zero address * I like this one

2) Tokens outside of zero address

The rational here is that a total supply reflecting the convention can be computed by

totalSupply - balanceOf(zeroAddress) if needed

on and off chain

> The methods `transfer` and `transferFrom` method MUST error when the balance of `from_` is insufficient.

Reviewer 4: `transfer` has no `from_` argument. (I get your point, but the language needs to be changed.)

Author: Here `from` is implied as the signer of the transaction

Author: See also
https://github.com/algorandfoundation/ARCs/pull/223/commits/a9174bc215e756fb12858db4c0657cc0b73d2527

> The `transferFrom` method MUST error unless called by an approved spender as defined by an extension defined in this ARC.

Reviewer 4: Why does this say "approved spender as defined by an extension defined in this ARC"? I think you're just referring to the `approve` system, right?

Author: Yes, the `approve` system allows for multiple operators to spend on the behalf of other account if approved. I supposed we can revise it as follows.

approved spender -> `spender`

This may be more understandable since it matches the usage in the surrounding language and ABI spec.

Author: Revision here
https://github.com/algorandfoundation/ARCs/pull/223/commits/a9174bc215e756fb12858db4c0657cc0b73d2527

Reworded as `spender` approved by `owner` instead of approved spender. 
Added note that `from` is `owner` in case of `transfer` method

> The methods `transfer` and `transferFrom` MUST emit a `Transfer` event that a transfer is successful.

Reviewer 4: ```suggestion
The methods `transfer` and `transferFrom` MUST emit a `Transfer` event.
```

> A `Transfer` event SHOULD be emitted, with `from_` being the zero address, when a token is first minted.

Reviewer 4: ```suggestion
A `Transfer` event SHOULD be emitted, with `from_` being the zero address, when a token is minted.
```

A `Transfer` event SHOULD be emitted, with `to_` being the zero address, when a token is destroyed.

> The `Approval` event MUST be emitted when the `approve` method is called successfully.

Reviewer 4: Do you want to "fix" the Ethereum misfeature here, and put `increaseAllowance` and `decreaseAllowance` into the standard?
https://www.adrianhetman.com/unboxing-erc20-approve-issues/

Author: I may want to fix this issue from the get-go because the ARC-200 spec implementation may easily be ported to a EVM based chain with this fix already applied. I haven't heard of any issues with front running on Algorand yet but this seems like a great way to future-proof the spec. Will be considering.

Author: @jannotti I am on the side of not adding `increaseAllowance` and `decreaseAllowance` to the standard for simplicity sakes. I do think that implementations may add non-standard methods similar to the way OpenZeppelin does it with ERC-20 (https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol)

I have a working ARC200 implementation that methods like  `increaseAllowance` and `decreaseAllowance` for convenience.

A value of zero for the `approve` method and the `Approval` event indicates no approval.

> When a `Transfer` event emits following the `transferFrom` method, this may also indicate that the approved value for the token is decremented.

Reviewer 4: It seems this should be stronger.  The `transferFrom` method MUST decrement the current approval value for the `from` address of the `to` address's tokens.

I don't know if you want to require that a new `Approval` event is emitted.

Author: This one could be revised to be more clear.

The approval event must be emitted in case of `approve` and `transferFrom`

`transferFrom` emits both `Approval` and `Transfer`

`approve` emits only `Approval`

Going to come back to this one. I see your point. Thx!

Author: Reworded here
https://github.com/algorandfoundation/ARCs/pull/223/commits/125f2342c4ab3bbb3e73918223811840d2fdfb5e

> The contract MUST allow multiple operators per owner.

Reviewer 4: I think this means that, for example, owner A can approval both account's B and C simultaneously.  Is that right?  Does the MUST here imply an unbounded number of such approvals must be allowed?

Author: Yes, there is no limit on approvals for any account.

For example,

approval[A][B] and
approval[A][C] 

are mutually exclusive.

On another note

approval[D][B] and
approval[D][C] 

MUST be allowed

that is there are no bounds on approval relations.

---

> I think this means that, for example, owner A can have an outstanding approval for both account's B and C to spend their tokens simultaneously. Is that right?

Yes, if A has approval for both B and C, B and C can be spent by A simultaneously.

---

Hope that helped

All methods in this standard that are marked as `readonly` MUST be read-only as defined by [ARC-22](./arc-0022.md).

## Rationale

This specification is based on <a href="https://eips.ethereum.org/EIPS/eip-20">ERC-20</a>.

### Core Specification

The core specification identical to ERC-20 with the exception of:

> - Method and event arguments are appended, rather than prefixed, with \_.

Reviewer 4: I don't think there's any reason to do this in an ARC.  There is no semantic meaning to the variable names here - if some HLL compiled the json down to something that could be called from the HLL and use the argument names from the standard, it ought to be the convention of the language in question how to handle name conflicts.

Author: I agree. It doesn't matter because when it comes does to it names are not translated in to the ABI where it counts. I'm going to revert this to the original state, ie. _from _to just like the ERC-20 spec.

Reviewer: What's the reason for `_` at all? (outside of aligning with ERC20)

Author: Yes. I will exterminate the tick ; )
